clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -776,32 +778,23 @@ public PartitionRecords(long fetchOffset, TopicPartition partition, List<Consume
         }
 
         private boolean isEmpty() {
-            return records == null || records.isEmpty();
+            return records == null || position >= records.size();
         }
 
         private void discard() {
             this.records = null;
         }
 
         private List<ConsumerRecord<K, V>> take(int n) {
-            if (records == null)
-                return new ArrayList<>();
+            if (isEmpty())
+                return Collections.emptyList();
 
-            if (n >= records.size()) {
-                List<ConsumerRecord<K, V>> res = this.records;
-                this.records = null;
-                return res;
-            }
-
-            List<ConsumerRecord<K, V>> res = new ArrayList<>(n);
-            Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
-            for (int i = 0; i < n; i++) {
-                res.add(iterator.next());
-                iterator.remove();
```

Comment:
Thinking about it, this may be the root of the problem. We are removing from the head of the ArrayList, one at a time. For each removal, we cause all the elements to be shifted in the underlying array, which is very inefficient if n is not small.